### PR TITLE
Add Barber Field Encoding for HttpUrl

### PR DESCRIPTION
**Pull request by adrw**
DocumentTemplate fields that are URLs do not need to be as wide open as plaintext but can't be as strict as full HTML escaping. Use of OkHttp.HttpUrl for parsing validation and escaping is a good approach.

### CODE REVIEW

File: barber/src/main/kotlin/app/cash/barber/BarberHttpUrlMustacheFactory.kt [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import okhttp3.HttpUrl.Companion.toHttpUrl
+import java.io.Writer
+
+/**
+ * Mustache Factory that handles [STRING_HTTPURL] fields and doesn't apply the HTML escaping
+ *  present in [DefaultMustacheFactory]
+ */
+class BarberHttpUrlMustacheFactory : DefaultMustacheFactory() {
+  override fun encode(value: String?, writer: Writer?) {
+    writer?.write(value?.toHttpUrl().toString()) ?: throw MustacheException("Null Writer. Failed to encode value: $value")
```
Comment (acostama) [inline barber/src/main/kotlin/app/cash/barber/BarberHttpUrlMustacheFactory.kt:15]: Safe to pass null to the writer?

File: barber/src/test/kotlin/app/cash/barber/BarberTest.kt [hunk 2]
```diff
@@ -525,44 +586,50 @@ class BarberTest {
...
-    assertEquals("""
+    assertEquals(
+      """
       |Errors
       |1) Unable to resolve compatible DocumentTemplate [version=4] for [templateToken=multiVersionDocumentTargetChange]
       |[compatibleOptions=[3]]
       |
-    """.trimMargin(), incompatibleVersionException.toString())
+    """.trimMargin(), incompatibleVersionException.toString()
+    )
```
Comment (acostama) [inline barber/src/test/kotlin/app/cash/barber/BarberTest.kt:609]: Looks like only formatting in this file?

File: barber/src/test/kotlin/app/cash/barber/BarbershopBuilderTest.kt [hunk 3]
```diff
@@ -352,54 +380,93 @@ class BarbershopBuilderTest {
...
+      EncodingTestDocument(
+        // Using the default HTML encoding, the apostrophe in no_annotation_field have been escaped.
+        no_annotation_field = "You purchased 100 shares of McDonald's.",
+        default_field = "You purchased 100 shares of McDonald&#39;s.",
+        html_field = "You purchased 100 shares of McDonald&#39;s.",
+        plaintext_field = "You purchased 100 shares of McDonald's.",
+        url_field = "https://test.com/%3Ciframe%3Etest%3C/iframe%3E%3Cstyle%3E@keyframes%20x%7B%7D%3C/style%3E%3Cb%20style=%22animation-n%3Ch1%3Etest2%3C/h1%3Ejavascript:alert%601%60",
+      )
+    )
+    assertThat(spec.url_field).doesNotContain(iframeInjection)
```
Comment (acostama) [inline barber/src/test/kotlin/app/cash/barber/BarbershopBuilderTest.kt:435]: 🔥

**Review by acostama — APPROVED** [review]
Lgtm

**Review by acostama — APPROVED** [review]

File: barber/src/test/kotlin/app/cash/barber/BarberTest.kt (hunk 2 above)
Comment (adrw) [reply barber/src/test/kotlin/app/cash/barber/BarberTest.kt:609]: Yup

File: barber/src/main/kotlin/app/cash/barber/BarberHttpUrlMustacheFactory.kt (hunk 1 above)
Comment (adrw) [reply barber/src/main/kotlin/app/cash/barber/BarberHttpUrlMustacheFactory.kt:15]: Thanks, will fix

**Pull request merged**